ENH: Make numpy.array_api more portable

asmeurer opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

- Make `.device` return a special `CPU_DEVICE` object (not present in the namespace) instead of `"cpu"`. See https://github.com/data-apis/array-api/issues/626.
- Make dtype objects separate classes, rather than reusing NumPy dtype objects. Fixes https://github.com/numpy/numpy/issues/23883
- Fix `nonzero` to disallow 0-D inputs (this is not related to portability, but it was a small change so I figured I would just include it here).
- Update the array API compatibility document accordingly. 

These will make it harder to write non-portable code that successfully works with `numpy.array_api`.

asmeurer committed:
Replace device="cpu" with a special object in numpy.array_api

This way, it does not appear that "cpu" is a portable device object across
different array API compatible libraries. See
https://github.com/data-apis/array-api/issues/626.

asmeurer committed:
Use separate wrapped dtype objects in numpy.array_api

This way there is no ambiguity about the fact the non-portability of NumPy
dtype behavior, or the fact that NumPy dtypes are not necessarily allowed as
dtypes for non-NumPy array APIs.

Fixes #23883

asmeurer committed:
Disallow 0-dimensional arrays in numpy.array_api.nonzero

This is required by the standard, even though np.nonzero does not yet disallow
it.

asmeurer committed:
Update some items in the array_api compatibility document

asmeurer commented on numpy/array_api/_dtypes.py:
```diff
@@ -1,21 +1,42 @@
+    def __repr__(self):
+        return f"np.array_api.{self._np_dtype.name}"
+
+    def __eq__(self, other):
```
I'm thinking it might be a good idea to make `==` explicitly raise an exception when compared against a `numpy` dtype object. Right now `numpy.float32 == numpy.array_api.float32` returns `False`, but I feel like this could mask bugs. Is there a straightforward way to detect if `other` is a NumPy dtype object?

mhvk replied:
This is a very fly-by comment, but in python generally `==` is supposed never to fail.

asmeurer replied:
Yes, I'm not completely sure about it. I did set this to error when I was creating this PR and it definitely helped me to find all the places in the code that needed to be updated. 

My worry if is someone is incorrectly doing:

```py
if x.dtype == np.float32: # x is a numpy.array_api array
    ...
```

or

```py
if x.dtype == 'f4':
    ...
```

The code will just subtly start being wrong.

But then again, I don't know if that kind of error actually exists in the wild. It is already wrong for other libraries like PyTorch.

OTOH, like you say, `==` failing can be very annoying as it would make things like `in` not work properly. I think I'm leaning towards not making this change, but I'm interested to hear what others think.

seberg replied:
Notwithstanding that this should move eventually.  For *this* namespace it seems right to raise errors liberally, since I think by now it is accepted as a *testing* implementation:  If it works with this, it should work with anything.

mhvk replied:
How about a warning? For testing purposes, those can be turned into errors.

seberg replied:
I don't have an opinion, TBH.  Besides testing, I never saw much purpose to the namespace as is and in that context I think liberally raising is OK too.

OTOH, if you formalize the "this is for testing" part of the namespace, then it does seem quite logical to have a custom `UndefinedBehavior` warning.  That would make it easier to ignore it in tests if there is a reason to do so (not sure there is, but maybe it doesn't matter).

charris commented:
This could use a release note.

asmeurer committed:
Add a release notes entry

BvB93 commented on numpy/array_api/_typing.py:
```diff
@@ -50,7 +50,7 @@ class NestedSequence(Protocol[_T_co]):
     def __len__(self, /) -> int: ...
 
-Device = Literal["cpu"]
+Device = type(CPU_DEVICE)
```
Dynamically assigning a type like this isn't gona work I'm afraid. Probably want to go with an `Any` here.
```suggestion
Device = Any
```

asmeurer replied:
We do know the type https://github.com/numpy/numpy/pull/25370/files#diff-79e6f716d807a9f60474911f90c087a717802dd6e63c82cd11689b7e47e943d1R45. I just figured this would work.

asmeurer committed:
Fix some type hint aliases for numpy.array_api

rgommers reviewed (commented):
Thanks @asmeurer. This looks quite close, I found one issue when testing this on SciPy. Also the `==` warning/exception for dtype comparison needs to be finalized.

rgommers commented:
Resolving the "2 blank lines" part of the linter complaints would also be nice to take along.

asmeurer commented on numpy/array_api/_creation_functions.py:
```diff
@@ -68,6 +63,8 @@ def asarray(
         # to an object array. TODO: This won't handle large integers in lists.
         raise OverflowError("Integer out of bounds for array dtypes")
+    if dtype is not None:
+        dtype = dtype._np_dtype
```
Good catch. The test suite didn't catch this because it failed on the NotImplementedError.

asmeurer committed:
Fix bug in numpy.array_api.asarray with copy=True

asmeurer committed:
Make numpy.array_api dtypes issue a warning when compared against numpy dtypes

This is to prevent user error, since something like numpy.array_api.float32 ==
numpy.float32 gives False.

asmeurer committed:
Fix some linter issues

rgommers commented:
The `test_warning_calls` failures are real. 

And now that gh-25317 is merged, it'd be good to update this PR for device handling of `fftfreq`/`rfftfreq`.

asmeurer committed:
Add stacklevel to a warning

asmeurer committed:
Merge branch 'main' into array_api-portability

asmeurer committed:
Fix the array_api fft creation functions to use the custom CPU_DEVICE object

rgommers commented:
This LGTM now, but the added strictness is hard to deal with without the fix suggested in https://github.com/data-apis/array-api-compat/issues/77#issuecomment-1894095240. So ideally that would be fixed first before merging this.

asmeurer commented:
@rgommers I'm not following how that array-api-compat issue relates to this.

rgommers commented:
> @rgommers I'm not following how that array-api-compat issue relates to this.

I ran into that issue when testing this PR with SciPy. All my attempts at fixing up SciPy to be robust against changes in this PR are blocked until that issue is fixed. Basically, the addition of `numpy.ndarray.__array_namespace__` changed the behavior of `array_api_compat` - but that was still okay when this module had some flexibility. The new dtype objects added in this PR are very restrictive, so it's impossible to get away with mixing `numpy` and `numpy.array_api`, or using `dtype=bool` (the Python builtin `bool`) after this gets merged.

tl;dr I don't really want to leave SciPy (and scikit-learn) in a state where tons of tests fail with NumPy `main`.

asmeurer commented:
I guess there must be some complexity in the scipy implementation for this that to be related to this. I'm surprised you have array-api compatible code that's mixing numpy dtypes, since that already doesn't work for pytorch or cupy.

rgommers commented:
Now that https://github.com/scipy/scipy/pull/19885 is merged, let's give this a go too. Thanks @asmeurer & reviewers!

rgommers commented:
@asmeurer it looks like we're ready now for you to create the standalone version of `numpy.array_api`, so that we can make an announcement and switch over the known consumers and remove `numpy.array_api` before the 2.0.x branch point.
